### PR TITLE
Bulk CSV export for views-based searches

### LINKED ISSUE #7354: Bulk export support for search backend
## What?
The search backend (https://github.com/Graylog2/graylog2-server/blob/master/graylog2-server/src/main/java/org/graylog/plugins/views/search/engine/QueryEngine.java etc) does not support bulk exports right now.
Depending on the elasticsearch version and sorting requirements, the implementations might look a bit different.

## Why?
Exporting results as CSV or other formats (plain log view with format strings, etc) require iterating over large amounts of data, which cannot be retrieved with deep pagination.
The current workaround is #7227 but that is not a good solution due to how Stream selection works now.

Moreover, the API layer should be able to support chunked transport to avoid buffering large amounts of results in the server process, so separate QueryEngine support seems necessary.

## Notes
The implementation should support sorting the data, where possible.
The implementation does not have to support aggregations, as those are blocking on the elasticsearch side anyway, and are not even supported in "scroll" requests.

## Your Environment
* Graylog Version: 3.2.0
* Elasticsearch Version: 5.6/6.8

### PR DESCRIPTION
**Pull request by alex-konn**
This PR contains the essentials for the new search export as described in #7354.

The following requirements will be implemented in follow-up PRs:
- audit log
- api docs
- tests for search after requests with multiple Elasticsearch versions

Next steps:
- Use the massage table widget name as a title for the CSV export file
- Only show stream related fields in CSV export modal

### CODE REVIEW

**Comment by alex-konn** [comment]
> I did a simple export and that worked for me. slightly_smiling_face Great work! tada
> 
> Apart from the inline comments I have some generic ones:
> 
>     * [ ]  We are currently using our internal ES date format for the `timestamp` value in the export (e.g. `2015-01-01 01:00:00.000"`) That timestamp format is quite useless because it's non-standard and doesn't include the timezone. I talked to @kroepke and we think that we should output an ISO8601 timestamp instead. (e.g. `2015-01-01T01:00:00.000Z`) This is basically  a special decorator for the `timestamp` field if it's present. This doesn't format other fields that use the non-standard format, but at least for the regular timestamp we use a sane format then.
> 
>     * [ ]  As discussed offline, we can probably add the default values from `Defaults` to the `MessagesRequest` auto-value itself. That way a `MessagesRequest.builder().build()` always returns a working request without having to check for defaults in several places, we can get rid of all the `Optional`s and we can also use hibernate validations instead of a manual validation method.
> 
>     * [ ]  Do we need the `additionalQueryString()` in `MessagesRequest`?
> 
> 
> I only looked at the backend code. Someone else needs to review the UI code, please. smiley

Great feedback! Thanks! :) The timestamp format issue is an obvious one and I feel slightly embarrassed that I overlooked it all this time. :smile:
`MessageRequest`defaults will be overhauled and I'll double check about `additionalQueryString()`. I added it, because I saw that `SearchType#query()` is used for constructing queries in `ElasticsearchBackend` and assumed that's relevant for message tables in dashboard, but I should verify that.

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/export/MessagesRequest.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        @JsonProperty("query_string")
+        public abstract Builder queryString(BackendQuery queryString);
+
+        public abstract Builder additionalQueryString(BackendQuery queryString);
+
+        @JsonProperty("fields_in_order")
+        public abstract Builder fieldsInOrder(LinkedHashSet<String> fieldsInOrder);
+
+        public Builder fieldsInOrder(String... fieldsInOrder) {
+            LinkedHashSet<String> fields = Arrays.stream(fieldsInOrder).collect(Collectors.toCollection(LinkedHashSet::new));
```
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/export/MessagesRequest.java:99]: :+1:

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/export/es/ElasticsearchExportBackend.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+import static org.elasticsearch.index.query.QueryBuilders.boolQuery;
+import static org.elasticsearch.index.query.QueryBuilders.matchAllQuery;
+import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
+import static org.elasticsearch.index.query.QueryBuilders.termsQuery;
+
+public class ElasticsearchExportBackend implements ExportBackend {
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchExportBackend.class);
+
+    private static final int DEFAULT_CHUNK_SIZE = 1000;
+    private static final String TIEBREAKER_FIELD = "gl2_message_id";
```
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/export/es/ElasticsearchExportBackend.java:61]: :+1:

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/export/ChunkedRunner.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+
+public class ChunkedRunner {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ChunkedRunner.class);
+
+    private final ChunkedOutput<SimpleMessageChunk> output = new ChunkedOutput<>(SimpleMessageChunk.class);
+
+    public static ChunkedOutput<SimpleMessageChunk> runAsyncc(Consumer<Consumer<SimpleMessageChunk>> call) {
```
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/export/ChunkedRunner.java:35]: HOW DARE YOU DOUBT MY SPELLING!

will fix...

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/ElasticsearchQueryString.java [hunk 4]
```diff
@@ -30,12 +31,21 @@
...
+
     @Override
     public abstract String type();
 
     @JsonProperty
     public abstract String queryString();
 
+    @JsonIgnore
+    public boolean isEmpty() {
+        return queryString().isEmpty() || queryString().equals("*");
```
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/ElasticsearchQueryString.java:46]: :+1:

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/export/MessagesResult.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog.plugins.views.search.export;
+
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class MessagesResult {
```
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/export/MessagesResult.java:22]: Holy shit. All this time..

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/export/es/ElasticsearchExportBackend.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        return ssb;
+    }
+
+    @SuppressWarnings("OptionalGetWithoutIsPresent")
+    private QueryBuilder queryFrom(MessagesRequest request) {
+        ElasticsearchQueryString backendQuery = (ElasticsearchQueryString) request.queryString().get();
+
+        QueryBuilder query = backendQuery.isEmpty() ?
+                matchAllQuery() :
+                queryStringQuery(backendQuery.queryString());//.allowLeadingWildcard(allowLeadingWildcardSearches);
```
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/export/es/ElasticsearchExportBackend.java:143]: :grimacing:

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/export/MessagesRequest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (alex-konn) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/export/MessagesRequest.java:108]: We actually do need `additionalQueryString` for dashboard message tables. 
They have their own query string plus the optional one specified at the top of the dashboard ("global override").
I'm not happy with having that in `MessagesRequest` either (and the name sucks of course), but it was the least awkward solution I could think of. `MessagesRequest` doubles as the request body bean and parameter object for `ElasticsearchExportBackend`. I could introduce a different parameter object and have `additionalQueryString` only on that, but I'm not convinced it's worth the extra object and mapping overhead.

**Comment by alex-konn** [comment]
> > I did a simple export and that worked for me. slightly_smiling_face Great work! tada
> > Apart from the inline comments I have some generic ones:
> > ```
> > * [ ]  We are currently using our internal ES date format for the `timestamp` value in the export (e.g. `2015-01-01 01:00:00.000"`) That timestamp format is quite useless because it's non-standard and doesn't include the timezone. I talked to @kroepke and we think that we should output an ISO8601 timestamp instead. (e.g. `2015-01-01T01:00:00.000Z`) This is basically  a special decorator for the `timestamp` field if it's present. This doesn't format other fields that use the non-standard format, but at least for the regular timestamp we use a sane format then.
> > 
> > * [ ]  As discussed offline, we can probably add the default values from `Defaults` to the `MessagesRequest` auto-value itself. That way a `MessagesRequest.builder().build()` always returns a working request without having to check for defaults in several places, we can get rid of all the `Optional`s and we can also use hibernate validations instead of a manual validation method.
> > 
> > * [ ]  Do we need the `additionalQueryString()` in `MessagesRequest`?
> > ```
> > 
> > 
> > I only looked at the backend code. Someone else needs to review the UI code, please. smiley
> 
> Great feedback! Thanks! :) The timestamp format issue is an obvious one and I feel slightly embarrassed that I overlooked it all this time. smile
> `MessageRequest`defaults will be overhauled and I'll double check about `additionalQueryString()`. I added it, because I saw that `SearchType#query()` is used for constructing queries in `ElasticsearchBackend` and assumed that's relevant for message tables in dashboard, but I should verify that.

We actually do need `additionalQueryString` for dashboard message tables. 
They have their own query string plus the optional one specified at the top of the dashboard ("global override").
I'm not happy with having that in `MessagesRequest` either (and the name sucks of course), but it was the least awkward solution I could think of. `MessagesRequest` doubles as the request body bean and parameter object for `ElasticsearchExportBackend`. I could introduce a different parameter object and have `additionalQueryString` only on that, but I'm not convinced it's worth the extra object and mapping overhead.

**Comment by bernd** [comment]
@alex-konn 
> We actually do need `additionalQueryString` for dashboard message tables.
> They have their own query string plus the optional one specified at the top of the dashboard ("global override").
> I'm not happy with having that in `MessagesRequest` either (and the name sucks of course), but it was the least awkward solution I could think of. `MessagesRequest` doubles as the request body bean and parameter object for `ElasticsearchExportBackend`. I could introduce a different parameter object and have `additionalQueryString` only on that, but I'm not convinced it's worth the extra object and mapping overhead.

Thanks for the clarification! Is the second query an actual override or do we need to use **both** queries? Otherwise we could just have a `MessagesRequest#query()` and the caller decides which query string will be used when creating the `MessagesRequest` object.. The "regular" one or the "override".

If we actually need to use both queries in the ES request, basically combining the two queries, then we need both in the `MessagesRequest` of course.

**Comment by alex-konn** [comment]
> @alex-konn
> 
> > We actually do need `additionalQueryString` for dashboard message tables.
> > They have their own query string plus the optional one specified at the top of the dashboard ("global override").
> > I'm not happy with having that in `MessagesRequest` either (and the name sucks of course), but it was the least awkward solution I could think of. `MessagesRequest` doubles as the request body bean and parameter object for `ElasticsearchExportBackend`. I could introduce a different parameter object and have `additionalQueryString` only on that, but I'm not convinced it's worth the extra object and mapping overhead.
> 
> Thanks for the clarification! Is the second query an actual override or do we need to use **both** queries? Otherwise we could just have a `MessagesRequest#query()` and the caller decides which query string will be used when creating the `MessagesRequest` object.. The "regular" one or the "override".
> 
> If we actually need to use both queries in the ES request, basically combining the two queries, then we need both in the `MessagesRequest` of course.

"global override" is what it's called in the code. actually it's applied as an additional filter on the query.

File: graylog2-web-interface/src/views/components/Select.jsx [hunk 7]
```diff
@@ -115,24 +121,34 @@ const Select = ({ components, styles, ignoreCase = true, ignoreAccents = false,
     valueContainer,
   };
   const filterOption = createFilter({ ignoreCase, ignoreAccents });
+  const selectProps = {
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/components/Select.jsx:124]: Extracting the props makes it harder to see how the effective props look like. Instead we could leave the structure as before (with inline props next to the component) and do something like: `const Component = allowOptionCreation ? CreatableSelect : ReactSelect;` and render `<Component ...` later.

File: graylog2-web-interface/src/views/components/searchbar/csvexport/startDownload.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    searchType = view.getSearchTypeByWidgetId(selectedWidget.id);
+  } else {
+    const viewTitle = view.title || `Untitled ${ViewTypeLabel({ type: view.type, capitalize: true })}`;
+    filename = `${viewTitle}-${filename}`;
+  }
+
+  const filenameWithoutSpaces = StringUtils.replaceSpaces(filename, '-');
+
+  if (view.type === View.Type.Dashboard) {
+    return _exportOnDashboard(defaultExportPayload, searchType, view.search.id, filenameWithoutSpaces);
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/components/searchbar/csvexport/startDownload.js:54]: Should we move this and the corresponding call for a search to the strategies as well?

File: graylog2-web-interface/src/views/components/searchbar/csvexport/CSVExportSettings.jsx [hunk 9]
```diff
@@ -0,0 +1,107 @@
+// @flow strict
+import React from 'react';
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/components/searchbar/csvexport/CSVExportSettings.jsx:2]: We should import this as
```suggestion
import * as React from 'react';
```
So we do not get only the default export.

File: graylog2-web-interface/src/views/components/Select.jsx (hunk 7 above)
Comment (linuspahl) [reply graylog2-web-interface/src/views/components/Select.jsx]: Thanks, this is a very good suggestion.

**Pull request merged**